src/wallet/wallet.cpp

```diff
@@ -3847,6 +3847,11 @@ bool CWallet::MigrateToSQLite(bilingual_str& error)
     m_database->Close();
     fs::remove(db_path);
 
+    if (db_path.filename() == GetName()) {
```

Comment:
If you load a wallet from an external path (not the wallet directory), the wallet name is set to the external path.

Reply:
So this applies to wallets outside the datadir, but not to the original default wallet sitting in the root of the data dir (instead of /wallets)?

Reply:
For wallets that are not in a wallet dir, the db's filename will be the name of the wallet itself.

Reply:
In commit "wallet: Be able to migrate wallets that are not in a wallet dir" (e22a3a471bf6696cb9ac96f3f348030cb07aa8b2)

This logic seems unnecessarily confusing. I think there is no benefit to writing:

```c++
fs::path db_dir = db_path.parent_path();
if (db_path.filename() == GetName()) {
    db_dir = db_path
}
```

Instead of simply:

```c++
const fs::path db_dir = GetName();
```

It also looks like `db_dir` variable is used only once, so you could probably get rid of it and just write `GetName()` directly below.

Reply:
It can't use `GetName()` directly since that is just the name of the wallet, not the path to where we want it to be stored as `db_dir` is in this case.

Reply:
Sorry, I overlooked that in my suggestion. But I think this function can do the same thing as [`MakeWalletDatabase`](https://github.com/bitcoin/bitcoin/blob/c92fd638860c5b4477851fb3790bc8068f808d3e/src/wallet/wallet.cpp#L2892), [`VerifyWallets`](https://github.com/bitcoin/bitcoin/blob/c92fd638860c5b4477851fb3790bc8068f808d3e/src/wallet/load.cpp#L79), [`RestoreWallet`](https://github.com/bitcoin/bitcoin/blob/c92fd638860c5b4477851fb3790bc8068f808d3e/src/wallet/wallet.cpp#L446), [`MigrateLegacyToDescriptor`](https://github.com/bitcoin/bitcoin/blob/c92fd638860c5b4477851fb3790bc8068f808d3e/src/wallet/wallet.cpp#L4261), and [`ExecuteWalletToolFunc`](https://github.com/bitcoin/bitcoin/blob/c92fd638860c5b4477851fb3790bc8068f808d3e/src/wallet/wallettool.cpp#L139) and derive the wallet path from the wallet name in the normal way. The following change seems to work:

```diff
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -3843,21 +3843,18 @@ bool CWallet::MigrateToSQLite(bilingual_str& error)
 
     // Close this database and delete the file
     fs::path db_path = fs::PathFromString(m_database->Filename());
-    fs::path db_dir = db_path.parent_path();
     m_database->Close();
     fs::remove(db_path);
 
-    if (db_path.filename() == GetName()) {
-        // For wallet files that aren't in a wallet dir, we will want to make a wallet dir that has the same name
-        db_dir = db_path;
-    }
+    // Turn wallet name into an absolute path that can be passed to passed to MakeDatabase.
+    const fs::path wallet_path = fsbridge::AbsPathJoin(GetWalletDir(), fs::PathFromString(m_name));
 
     // Make new DB
     DatabaseOptions opts;
     opts.require_create = true;
     opts.require_format = DatabaseFormat::SQLITE;
     DatabaseStatus db_status;
-    std::unique_ptr<WalletDatabase> new_db = MakeDatabase(db_dir, opts, db_status, error);
+    std::unique_ptr<WalletDatabase> new_db = MakeDatabase(wallet_path, opts, db_status, error);
     assert(new_db); // This is to prevent doing anything further with this wallet. The original file was deleted, but a backup exists.
     m_database.reset();
     m_database = std::move(new_db);
```